File: applications/rtkoverlayphaseandshroud/rtkoverlayphaseandshroud.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkoverlayphaseandshroud_ggo.h"
#include "rtkMacro.h"

#include <itkImageFileReader.h>
#include <itkImageFileWriter.h>
#include <itkCSVArray2DFileReader.h>
#include <itkRGBPixel.h>
#include <itkImageRegionConstIterator.h>
#include <itkImageRegionIteratorWithIndex.h>
#include <itkNumericTraits.h>

#include <fstream>

int
main(int argc, char * argv[])
{
  GGO(rtkoverlayphaseandshroud, args_info);

  using InputPixelType = double;
  using RGBPixelType = itk::RGBPixel<unsigned char>;
  constexpr unsigned int Dimension = 2;

  using InputImageType = itk::Image<InputPixelType, Dimension>;
  using OutputImageType = itk::Image<RGBPixelType, Dimension>;

  // Read
  InputImageType::Pointer readImage;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(readImage = itk::ReadImage<InputImageType>(args_info.input_arg))

  // Read signal file
  using ReaderType = itk::CSVArray2DFileReader<double>;
  auto signalReader = ReaderType::New();
  signalReader->SetFileName(args_info.signal_arg);
  signalReader->SetFieldDelimiterCharacter(';');
  signalReader->HasRowHeadersOff();
  signalReader->HasColumnHeadersOff();
  TRY_AND_EXIT_ON_ITK_EXCEPTION(signalReader->Update())
  std::vector<double> signal = signalReader->GetArray2DDataObject()->GetColumn(0);

  // Locate the minima in the signal
  std::vector<bool> minima;
  minima.push_back(false);
  for (unsigned int i = 1; i < signal.size(); i++)
  {
    minima.push_back((signal[i] < signal[i - 1]));
  }

  // Create output RGB image
  auto RGBout = OutputImageType::New();
  RGBout->SetRegions(readImage->GetLargestPossibleRegion());
  RGBout->Allocate();

  // Compute min and max of shroud to scale output
  itk::ImageRegionConstIterator<InputImageType>      itIn(readImage, readImage->GetLargestPossibleRegion());
  itk::ImageRegionIteratorWithIndex<OutputImageType> itOut(RGBout, RGBout->GetLargestPossibleRegion());

  double min = itk::NumericTraits<double>::max();
  double max = -itk::NumericTraits<double>::max();
  while (!itIn.IsAtEnd())
  {
    double currentPixel = itIn.Get();
    if (currentPixel < min)
      min = currentPixel;
    if (currentPixel > max)
      max = currentPixel;
    ++itIn;
  }

  // Fill the output
  itIn.GoToBegin();
  itOut.GoToBegin();
  while (!itOut.IsAtEnd())
  {
    RGBPixelType pix;

    if (minima[itOut.GetIndex()[1]]) // If it is a minimum, draw a red pixel
    {
      pix.Fill(0);
      pix.SetRed(255);
    }
    else // Otherwise, copy the input image, scaled to [0 - 255], in all channels
    {
      pix.Fill(floor((itIn.Get() - min) * 255.0 / (max - min)));
    }

    itOut.Set(pix);

    ++itIn;
    ++itOut;
  }

  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(RGBout, args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkoverlayphaseandshroud/rtkoverlayphaseandshroud.py
#!/usr/bin/env python
import argparse
import itk
from itk import RTK as rtk
import numpy as np


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Generates an RGB image showing the phase peaks on top of the shroud"
    )

    parser.add_argument(
        "--input", "-i", help="Input shroud image file name", type=str, required=True
    )
    parser.add_argument(
        "--output",
        "-o",
        help="Output RGB image file name (.png, .jpg, ...)",
        type=str,
        required=True,
    )
    parser.add_argument(
        "--signal",
        help="File containing the phase of each projection",
        type=str,
        required=True,
    )

    return parser


def process(args_info: argparse.Namespace):
    if args_info.verbose:
        print(f"Reading input image: {args_info.input}")

    Dimension = 2
    InputImageType = itk.Image[itk.D, Dimension]
    OutputImageType = itk.Image[itk.RGBPixel[itk.UC], Dimension]

    # Read input image
    readImage = itk.imread(args_info.input, itk.D)

    # Read signal file
    signal = np.loadtxt(args_info.signal, delimiter=";", usecols=0)

    # Locate the minima in the signal
    minima = np.zeros(signal.shape, dtype=bool)
    if signal.size > 1:
        minima[1:] = signal[1:] < signal[:-1]

    # Compute min and max of shroud to scale output
    in_array = itk.array_from_image(readImage)
    min_val = float(in_array.min())
    max_val = float(in_array.max())

    # Create output RGB image
    RGBout = OutputImageType.New()
    RGBout.SetRegions(readImage.GetLargestPossibleRegion())
    RGBout.SetSpacing(readImage.GetSpacing())
    RGBout.SetOrigin(readImage.GetOrigin())
    RGBout.SetDirection(readImage.GetDirection())
    RGBout.Allocate()

    # Fill the output via a writable numpy view (shape: Y x X x 3)
    rgb_view = itk.array_view_from_image(RGBout)

    # Scale input to [0, 255] and copy to all channels
    if max_val == min_val:
        scaled = np.full(in_array.shape, 255, dtype=np.uint8)
    else:
        scaled = np.clip(
            np.floor((in_array - min_val) * 255.0 / (max_val - min_val)), 0, 255
        ).astype(np.uint8)

    rgb_view[:, :, 0] = scaled
    rgb_view[:, :, 1] = scaled
    rgb_view[:, :, 2] = scaled

    # If it is a minimum row, draw a red pixel
    rgb_view[minima, :, 0] = 255
    rgb_view[minima, :, 1] = 0
    rgb_view[minima, :, 2] = 0

    itk.imwrite(RGBout, args_info.output)


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
